src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGRStrategies.hpp

```diff
@@ -19,67 +19,518 @@
-                                                            arraySlice1d< localIndex const > const & numLocalDofsPerField )
+template< int NLEVEL >
+class MGRStrategyBase
 {
-  array1d< HYPRE_Int > ret;
-  HYPRE_Int numFields = LvArray::integerConversion< HYPRE_Int >( numLocalDofsPerField.size() );
+public:
+
+  static constexpr HYPRE_Int numLevels = NLEVEL;       ///< Number of levels
+
+protected:
+
+  HYPRE_Int m_numBlocks{ 0 };                          ///< Number of different matrix blocks treated separately
+
+  std::vector< HYPRE_Int > m_labels[numLevels]{};      ///< Dof labels kept at each level
+  HYPRE_Int m_numLabels[numLevels]{ -1 };              ///< Number of dof labels kept
+  HYPRE_Int * m_ptrLabels[numLevels]{ nullptr };       ///< Pointers to each level's labels, as consumed by MGR
+
+  HYPRE_Int m_levelFRelaxMethod[numLevels]{ -1 };      ///< F-relaxation method for each level
+  HYPRE_Int m_levelInterpType[numLevels]{ -1 };        ///< Interpolation type for each level
+  HYPRE_Int m_levelRestrictType[numLevels]{ -1 };      ///< Restriction type for each level
+  HYPRE_Int m_levelCoarseGridMethod[numLevels]{ -1 };  ///< Coarse grid method for each level
+
+  HYPRE_Int m_numRestrictSweeps{ -1 };                 ///< Number of restrict sweeps
+  HYPRE_Int m_numInterpSweeps{ -1 };                   ///< Number of interpolation sweeps
+
+  HYPRE_Int m_fRelaxMethod{ -1 };                      ///< F-relaxation method
+  HYPRE_Int m_relaxType{ -1 };                         ///< F-relaxation type
+  HYPRE_Int m_numRelaxSweeps{ -1 };                    ///< F-relaxation number of sweeps
 
-  if( numFields > 0 )
+  HYPRE_Int m_globalSmoothType{ -1 };                  ///< Global smoothing type
+  HYPRE_Int m_numGlobalSmoothSweeps{ -1 };             ///< Global smoothing number of iterations
+
+  /**
+   * @brief Constructor.
+   * @param numBlocks number of blocks
+   */
+  explicit MGRStrategyBase( HYPRE_Int const numBlocks )
+    : m_numBlocks( numBlocks )
+  {}
+
+  /**
+   * @brief Call this after populating lv_cindexes.
+   */
+  void setupLabels()
   {
-    HYPRE_Int numTotalLocalDof = 0;
-    for( HYPRE_Int i = 0; i < numFields; ++i )
+    for( HYPRE_Int i = 0; i < numLevels; ++i )
     {
-      numTotalLocalDof += LvArray::integerConversion< HYPRE_Int >( numLocalDofsPerField[i] );
+      m_numLabels[i] = m_labels[i].size();
+      m_ptrLabels[i] = m_labels[i].data();
     }
+  }
+};
+
+/**
+ * @brief SinglePhasePoroelastic strategy.
+ *
+ * dofLabel: 0 = displacement, x-component
+ * dofLabel: 1 = displacement, y-component
+ * dofLabel: 2 = displacement, z-component
+ * dofLabel: 3 = pressure
+ *
+ * Ingredients:
+ * 1. F-points displacement (0,1,2), C-points pressure (3)
+ * 2. F-points smoother: AMG, single V-cycle, separate displacement components
+ * 3. C-points coarse-grid/Schur complement solver: boomer AMG
+ * 4. Global smoother: none
+ */
+class SinglePhasePoroelastic : public MGRStrategyBase< 1 >
```

Comment:
@castelletto1 @francoishamon I refactored MGR recipes into these classes. The class constructor chooses all relevant MGR parameter values, and then the `setup` function applies them on the solver. Ideally I wanted the recipes to be purely static sets of parameter choices which then get applied by some generic common code, but some dynamism is necessary (in particular, to set up coarse indices in variable-size problems like multiphase; but also maybe some parameter choices could depend on runtime characteristics of the problem, like size?), so I left it like this. We can feed more data about the problem into the constructor as we need it.

Anyway, I *really* hope that I didn't mess up when transferring the parameter values. I'm running a CO2 tutorial and the two multiphase benchmarks in the repo that use MGR, but for other solvers I will have to just switch some of the integrated tests to MGR to verify. But even then I don't know what optimal performance looked like. Let me know if you see anything weird here of if you have other cases you'd like me to run.

Reply:
On a side note, it baffles me that hypre does not provide named constants for various parameter choices, at least in the form of macro defs, and keeps using raw integer constants - even in its central, most established components like BoomerAMG. If you're lucky, you'll find a description in function comment, but I found those to be missing or out-of-date sometimes.

If hypre has as fundamental/philosophical reason to not provide constants, do you think it would be worthwhile defining them on the GEOSX side? E.g.
```c++
namespace geosx { 
namespace hypre {
enum RelaxType : HYPRE_Int
{
  jacobi = 0, // Jacobi
  seqGaussSeidel = 1, // Gauss-Seidel, sequential (very slow!)
  ...
  l1jacobi = 18 // \f$\ell_1\f$-scaled jacobi
};
} }
```